File: 3-leetcode/02-medium/79-word-search.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
  bool exist(vector<vector<char>> &board, string word)
  {
    int r = board.size();
    int c = board[0].size();
    for (unsigned int i = 0; i < r; i++)
      for (unsigned int j = 0; j < c; j++)
        if (dfs(board, word, i, j))
          return true;
    return false;
  }

  bool dfs(vector<vector<char>> &board, string &word, int r, int c)
  {
    if (!word.size())
      return true;
    if (r < 0 || r >= board.size() || c < 0 || c >= board[0].size() || board[r][c] != word[0])
      return false;
    char curr = board[r][c];
    board[r][c] = '*';
    string s = word.substr(1);
    bool ret = dfs(board, s, r - 1, c) || dfs(board, s, r + 1, c) || dfs(board, s, r, c - 1) || dfs(board, s, r, c + 1);
    board[r][c] = curr;
    return ret;
  }
};

File: 3-leetcode/02-medium/js/79-word-search.js
const exist = function (board, word) {
  const m = board.length;
  const n = board[0].length;
  const len = word.length - 1;
  const dirs = [
    [-1, 0],
    [0, 1],
    [1, 0],
    [0, -1],
  ];
  let flag = false;
  let visited = Array.from({ length: m }, (v) =>
    Array.from({ length: n }, (v) => false),
  );

  for (let i = 0; i < m && !flag; i++) {
    for (let j = 0; j < n && !flag; j++) {
      DFS(0, '', i, j);
    }
  }

  function DFS(level, str, r, c) {
    if (flag) return;
    if (board[r][c] !== word[level]) return;
    if (level === len) return (flag = true);
    str += board[r][c];
    visited[r][c] = true;
    for (let dir of dirs) {
      const [nr, nc] = [r + dir[0], c + dir[1]];
      if (
        nr >= 0 &&
        nr < m &&
        nc >= 0 &&
        nc < n &&
        board[nr][nc] === word[level + 1] &&
        !visited[nr][nc]
      ) {
        DFS(level + 1, str, nr, nc);
      }
    }
    visited[r][c] = false;
  }

  return flag;
};
